install_subrepos.py

```diff
@@ -0,0 +1,140 @@
+        continue
+    try:
+        dist = distribution(p.name)._path
+    except PackageNotFoundError:
+        dist = None
+        editable = None
+    else:
+        editable = (p == dist or p in dist.parents)
+
+    REPOS[p.name] = {'repo': p, 'dist': dist, 'editable': editable}
+
+# ---- Setup logger
+fmt = Formatter('%(asctime)s [%(levelname)s] [%(name)s] -> %(message)s')
+h = StreamHandler()
+h.setFormatter(fmt)
+logger = getLogger('InstallSubRepos')
+logger.addHandler(h)
+logger.setLevel('INFO')
+
+
+def get_python_lsp_version():
+    """Get current version to pass it to setuptools-scm."""
+    req_file = DEVPATH / 'requirements' / 'conda.txt'
+    with open(req_file, 'r', encoding='utf-8') as f:
+        for line in f:
+            if 'python-lsp-server' not in line:
+                continue
+
+            version = next(iter(Requirement(line).specifier)).version
+
+            break
+
+    return version
+
+
+def install_repo(name, editable=False):
+    """
+    Install a single repo from source located in spyder/external-deps, ignoring
+    dependencies, in standard or editable mode.
+
+    Parameters
+    ----------
+    name : str
+        Must be the distribution name of a repo in spyder/external-deps.
+    editable : bool (False)
+        Standard install (False) or editable (True). This uses the `-e` flag.
+
+    """
+    try:
+        repo_path = REPOS[name]['repo']
+    except KeyError:
+        logger.warning(
+            'Distribution %r not valid. Must be one of %s', name, set(REPOS.keys()))
+        return
+
+    install_cmd = BASE_COMMAND.copy()
+
+    # PyLSP requires pretend version
+    env = None
+    if name == 'python-lsp-server':
+        env = {**os.environ}
+        env.update({'SETUPTOOLS_SCM_PRETEND_VERSION': get_python_lsp_version()})
+
+    if editable:
+        # Add edit flag to install command
+        install_cmd.append('-e')
+        mode = 'editable'
+    else:
+        mode = 'standard'
+
+    logger.info('Installing %r from source in %s mode.', name, mode)
+    install_cmd.append(repo_path.as_posix())
+    check_output(install_cmd, env=env)
+
+
+def main(install=tuple(REPOS.keys()), **kwargs):
+    """
+    Install all subrepos from source.
+
+    Parameters
+    ----------
+    install : iterable (all repos in spyder/external-deps)
+        Distribution names of repos to be installed from spyder/external-deps.
+    **kwargs :
+        Keyword arguments passed to `install_repo`.
+
+    """
+    for repo in install:
+        install_repo(repo, **kwargs)
+
+
+if __name__ == '__main__':
+    # ---- Parse command line
+
+    parser = argparse.ArgumentParser(
+        usage="python install_subrepos.py [options]")
+    parser.add_argument(
+        '--install', dest='install', nargs='+',
+        default=REPOS.keys(),
+        help="Space-separated list of distribution names to install, e.g. "
+             "spyder spyder-kernels. If option not provided, then all of the "
```

Comment:
Dependeing on the comment above `spyder` should be removed

Reply:
@dalthviz, thanks for the feedback. I'll take a look as soon as possible.

Reply:
@dalthviz, I've fixed the offending documentation and instructions in the contributing guide. Yes, `spyder` should be installed in editable mode from the local source. The whole workflow should look like
```bash
$ conda create -n spyder-dev python=3
$ conda activate spyder-dev
(spyder-dev) $ conda install -c conda-forge --file requirements/conda.txt
(spyder-dev) $ pip install --no-deps .
(spyder-dev) $ python install_subrepos.py --editable
(spyder-dev) $ python bootstrap.py
```

Reply:
Some issues with the above: 

* This creates a `defaults` environment, but then installs Spyder from `conda-forge`, which ~crosses the streams~ mixes channels unnecessarily and could cause issues, especially down the road.
* To avoid both this problem and an extra step (and lengthy solve), you can just do ``conda create -n spyder-dev --file requirements/conda.txt`` and skip the third step
* `pip install --no-deps .` installs Spyder in non-editable mode, which is almost certainly not what you want and doesn't actually seem to accomplish anything. At the very least, you should do `pip install --no-deps -e .` to actually get an editable install
* Since it would save another extra step and essentially performs the same process, why not add `spyder` itself to the list of repos to install in the `install_subrepo.py` script, with a ``--no-spyder`` option to skip it?
* Just to note, with the improvements made here, its actually no longer strictly necessary to run `python bootstrap.py` if you're just going to run it without options, as you _should_ be able to just run `spyder` from anywhere and get the dev version, so long as your environment is activated. Its probably still worth recommending, since it runs some extra checks with helpful feedback, but experienced devs might want to consider it for routine use.

So, overall, the workflow I recommend is:

```shell
conda create -n spyder-dev -c conda-forge --file requirements/conda.txt
conda activate spyder-dev
pip install --no-deps -e .
python install_subrepos.py --editable
python -b -X dev bootstrap.py
```

Or, for core devs in routine usage, if the standard `spyder` entry point is used and `pip install --no-deps -e .` is incorporated into the `install_subrepos.py` script, this can reduce to

```shell
conda create -n spyder-dev -c conda-forge --file requirements/conda.txt
conda activate spyder-dev
python install_subrepos.py --editable
spyder
```

This keeps things as simple as the status quo workflow, which for reference is:

```shell
conda create -n spyder-dev python=3
conda activate spyder-dev
conda install -c conda-forge --file requirements/conda.txt
python bootstrap.py
```

I'll make review comments to this effect.

Reply:
@CAM-Gerlach, you are absolutely correct. The environment and requirements file should be on the same line to avoid channel mixing, and spyder should be installed in editable mode. On both accounts, I was being lazy here; my apologies. I will update the contributing file to reflect that. Regarding including spyder in install_subrepos.py, I think it was in there before and somehow was removed. I think it would be a great idea to move it back in and rename the module appropriately. I don't know that we need a `--no-spyder` flag, since one can install any custom combination by providing a list of what is wanted. However, I don't feel too strongly about this; if there are certain options that make it easier for most developers, then I'm in favor of it.

Reply:
It seems that when installed via `pip install -e .`, the `spyder` entrypoint runs up against what appears to the same requirement incompatibility issue that happens when testing Windows installers in CI (whereas the script wrappers installed without `-e` work just fine, as does `python -m spyder.app.start` in editable mode). Here's the full traceback when run as a `console_script` (to be able to see it, otherwise `spyder` doesn't print anything):

<details>

<summary>Full traceback</summary>

```python-traceback
Traceback (most recent call last):
  File "C:\Miniconda3\envs\spyder-dev\lib\site-packages\pkg_resources\__init__.py", line 581, in _build_master
    ws.require(__requires__)
  File "C:\Miniconda3\envs\spyder-dev\lib\site-packages\pkg_resources\__init__.py", line 909, in require
    needed = self.resolve(parse_requirements(requirements))
  File "C:\Miniconda3\envs\spyder-dev\lib\site-packages\pkg_resources\__init__.py", line 800, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (spyder-kernels 2.4.0.dev0 (c:\users\c. a. m. gerlach\documents\dev\spyderdev\spyder\external-deps\spyder-kernels), Requirement.parse('spyder-kernels<2.4.0,>=2.3.0'), {'spyder'})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Miniconda3\envs\spyder-dev\Scripts\spyder-script.py", line 33, in <module>
    sys.exit(load_entry_point('spyder', 'console_scripts', 'spyder')())
  File "C:\Miniconda3\envs\spyder-dev\Scripts\spyder-script.py", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "C:\Miniconda3\envs\spyder-dev\lib\importlib\metadata.py", line 86, in load
    module = import_module(match.group('module'))
  File "C:\Miniconda3\envs\spyder-dev\lib\importlib\__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "c:\users\c. a. m. gerlach\documents\dev\spyderdev\spyder\spyder\app\start.py", line 47, in <module>
    from spyder.utils.external import lockfile
  File "c:\users\c. a. m. gerlach\documents\dev\spyderdev\spyder\spyder\utils\external\lockfile.py", line 31, in <module>
    from spyder.utils.programs import is_spyder_process
  File "c:\users\c. a. m. gerlach\documents\dev\spyderdev\spyder\spyder\utils\programs.py", line 28, in <module>
    import pkg_resources
  File "C:\Miniconda3\envs\spyder-dev\lib\site-packages\pkg_resources\__init__.py", line 3260, in <module>
    def _initialize_master_working_set():
  File "C:\Miniconda3\envs\spyder-dev\lib\site-packages\pkg_resources\__init__.py", line 3234, in _call_aside
    f(*args, **kwargs)
  File "C:\Miniconda3\envs\spyder-dev\lib\site-packages\pkg_resources\__init__.py", line 3272, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "C:\Miniconda3\envs\spyder-dev\lib\site-packages\pkg_resources\__init__.py", line 583, in _build_master
    return cls._build_from_requirements(__requires__)
  File "C:\Miniconda3\envs\spyder-dev\lib\site-packages\pkg_resources\__init__.py", line 596, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "C:\Miniconda3\envs\spyder-dev\lib\site-packages\pkg_resources\__init__.py", line 795, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'spyder-kernels<2.4.0,>=2.3.0' distribution was not found and is required by spyder
```

The problem seems to have to do specifically with environment consistency checking when running via the script wrapper in editable mode; I'd have to dig deeper to find the exact reason why it triggers in this circumstance but not others, but it can be avoided by properly setting the requirement specifiers in `setup.py` to allow the current development versions (since they _are_ actually supported by the `version` given), or by extending the existing conditional requirements loosening for Windows installers on PRs to cover all `.dev` installs, and include the other deps too; i.e. changing the following in `setup.py`

```python
# Replace spyder-kernels constraint to enable
# building Windows installers on PRs
if 'dev' in __version__ and WINDOWS_INSTALLER_NAME:
    install_requires.remove('spyder-kernels>=2.3.0,<2.4.0')
    install_requires.append('spyder-kernels>=2.3.0,<=2.4.0.dev0')
```

to 

```python
# Loosen constraints to ensure dev versions still work
if 'dev' in __version__:
    install_requires.remove('python-lsp-server[all]>=1.4.1,<1.5.0')
    install_requires.remove('qtconsole>=5.3.0,<5.4.0')
    install_requires.remove('spyder-kernels>=2.3.0,<2.4.0')
    install_requires.append('python-lsp-server[all]>=1.4.1,<1.6.0')
    install_requires.append('qtconsole>=5.3.0,<5.5.0')
    install_requires.append('spyder-kernels>=2.3.0,<2.5.0')
```